File: gemma_001/retry_script.py
import time
from typing import Callable, Any, Type

def retry_on_failure(
    func: Callable[..., Any],
    max_attempts: int = 3,
    delay_seconds: float = 1.0,
    *args,
    exception_types: tuple[Type[Exception], ...] = (Exception,)
) -> Any:
    """
    Retries the execution of a function upon failure.

    Args:
        func: The function to execute.
        max_attempts: The maximum number of times to attempt the function call.
        delay_seconds: The time to wait between attempts in seconds.
        *args: Positional arguments to pass to the function.
        exception_types: A tuple of exception types to catch and retry on.
                          Defaults to catching any Exception.

    Returns:
        The result of the function if successful.

    Raises:
        The last encountered exception if all attempts fail.
    """
    last_exception = None
    for attempt in range(max_attempts):
        try:
            print(f"Attempt {attempt + 1}/{max_attempts}...")
            return func(*args)
        except exception_types as e:
            last_exception = e
            if attempt < max_attempts - 1:
                print(f"Attempt failed with {type(e).__name__}: {e}. Retrying in {delay_seconds} seconds...")
                time.sleep(delay_seconds)
            else:
                print("All attempts failed.")
                raise last_exception
        except Exception as e:
            # Catch any other unexpected exception and stop trying
            print(f"Caught unexpected exception {type(e).__name__} on attempt {attempt + 1}. Stopping retries.")
            raise e

# Example Usage (Self-contained for demonstration)
def unreliable_service_call():
    if not hasattr(unreliable_service_call, 'count'):
        unreliable_service_call.count = 0
    unreliable_service_call.count += 1
    if unreliable_service_call.count < 3:
        raise ConnectionError("Service unavailable")
    return "Success after retries"

if __name__ == '__main__':
    try:
        result = retry_on_failure(
            unreliable_service_call,
            max_attempts=5,
            delay_seconds=0.5,
            exception_types=(ConnectionError,)
        )
        print(f"Final Result: {result}")
    except Exception as e:
        print(f"Operation failed permanently: {type(e).__name__}: {e}")

File: gemma_001/retry_script.ts
/**
 * Executes an asynchronous function, retrying upon failure.
 *
 * @template T The return type of the successful function call.
 * @template A The type of the function arguments.
 * @param {async (...args: A) => Promise<T>} asyncFunc The async function to execute.
 * @param {number} [maxAttempts=3] The maximum number of times to attempt the call.
 * @param {number} [delayMs=1000] The delay in milliseconds between attempts.
 * @param {...A} args Arguments to pass to the async function.
 * @returns {Promise<T>} A promise that resolves with the function's result or rejects with the last error.
 */
async function retryOnFailure<T, A extends any[]>(
    asyncFunc: (...args: A) => Promise<T>,
    maxAttempts: number = 3,
    delayMs: number = 1000,
    ...args: A
): Promise<T> {
    let lastError: Error | any = null;

    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
        try {
            console.log(`Attempt ${attempt}/${maxAttempts}...`);
            // Await the result of the asynchronous function
            return await asyncFunc(...args);
        } catch (error) {
            lastError = error instanceof Error ? error : new Error(String(error));

            if (attempt < maxAttempts) {
                console.warn(`Attempt failed: ${lastError.message}. Retrying in ${delayMs / 1000} seconds...`);
                // Wait before the next attempt
                await new Promise(resolve => setTimeout(resolve, delayMs));
            } else {
                console.error("All attempts failed.");
                // Re-throw the last error after the loop finishes
                throw lastError;
            }
        }
    }
    // This line is technically unreachable if maxAttempts >= 1, but satisfies typescript compiler
    throw lastError || new Error("Unknown failure"); 
}

// Example Usage (Self-contained for demonstration)
let callCountTS = 0;
const unreliableServiceCallTS = async (...args: any[]): Promise<string> => {
    callCountTS++;
    if (callCountTS < 3) {
        throw new Error("Service unavailable");
    }
    return "Success after retries";
};

(async () => {
    try {
        const result = await retryOnFailure(unreliableServiceCallTS, 5, 500);
        console.log(`\nFinal Result: ${result}`);
    } catch (e) {
        console.error(`\nOperation failed permanently: ${(e as Error).message || e}`);
    }
})();